File: static/game/GameClient/Objects/Figures/Figure/Figure.js
import {GameObject} from "../../GameObject/GameObject.js";

export class Figure extends GameObject{
    constructor(name, sprite_atlas_url, isEnemy) {
        super(name, sprite_atlas_url);
        this.couldBeMoved = false;
        this.attack_radius = 2;
        this.move_radius = 2;
        this.health = "100";
        this.damage = 25;
        this.isEnemy = isEnemy;
        this.showHealth = true;
    }
    
    setProps(props = {
        'move_radius': 0,
        'attack_radius': 0,
        'health': 0,
        'damage': 0
    }){
        this.move_radius = props['move_radius'];
        this.attack_radius = props['attack_radius'];
        this.setHealth(props['health']);
        this.damage = props['damage'];
    }

    isDead() {
        let health = Number(this.health);
        return health <= 0;
    }

    setHealth(hp){
        if(hp < 0)
            hp = 0
        let health = String(hp);
        for(let i = 0; i < 4 - health.length; ++i){
            health = "0" + health;
        }
        if(Number(health) > 100)health = "100";
        if(Number(health) < 0)health = "000";
        this.health = health;
    }

    getDamage(hp){
        this.setHealth(Number(this.health) - hp);
    }

    addHealth(hp){
        this.setHealth(Number(this.health) + hp);
    }

    render(context) {
        super.render(context);
        if(this.showHealth){
            if(this.isEnemy){
                context.fillStyle = "rgba(255, 100, 100, 0.8)";
            }else{
                context.fillStyle = "rgba(0, 255, 0, 0.8)";
            }
            context.strokeStyle = "black";
            context.fillRect(this.rect.x + this.rect.width / 4, this.rect.y + this.rect.height * 0.75, this.rect.width - this.rect.width / 2, this.rect.height / 4);
            context.strokeText(this.health, this.rect.x + this.rect.width / 2.5, this.rect.y + this.rect.height * 0.75 + this.rect.height / 6);
        }
    }

    getPositionReport(){
        return {
            'id': this.id,
            'name': this.name,
            'x': this.x,
            'y': this.y
        }
    }
}

File: game/GameServer/Objects/Figure/Figure.py
from ..GameObject.GameObject import GameObject


class Figure(GameObject):
    def __init__(self, x: int, y: int, master: str, name: str):
        super(Figure, self).__init__(x, y)
        self.name = name
        self.health = 100
        self.move_radius = 2
        self.attack_radius = 2
        self.damage = 25
        self.master = master

    def isDead(self):
        return self.health <= 0

    def setHealth(self, hp: int):
        self.health = 100 if hp > 100 else 0 if hp < 0 else hp

    def addHealth(self, hp: int):
        self.setHealth(self.health + hp)

    def getDamage(self, hp: int):
        self.setHealth(self.health - hp)

    def getReport(self):
        report = {
            'name': self.name,
            'master': self.master,
            'x': self.x,
            'y': self.y,
            'props': {
                'move_radius': self.move_radius,
                'attack_radius': self.attack_radius,
                'health': self.health,
                'damage': self.damage
            }
        }
        return report
